### PR TITLE
Refactor hessian service and torch sample layer attention. 

### PR DESCRIPTION
**Pull request by irenaby**
## Pull Request Description:
Refactor hessian service.
Add data util for torch.
Update GPTQ trainer - use new data loader for samples and corresponding loss weights, rewrite SLA. 

## Checklist before requesting a review:
- [ ] I set the appropriate labels on the pull request.
- [ ] I have added/updated the release note draft (if necessary).
- [ ] I have updated the documentation to reflect my changes (if necessary).
- [ ] All function and files are well documented. 
- [ ] All function and classes have type hints.
- [ ] There is a licenses in all file.
- [ ] The function and variable names are informative. 
- [ ] I have checked for code duplications.
- [ ] I have added new unittest (if necessary).

### CODE REVIEW

File: model_compression_toolkit/core/pytorch/data_util.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
+# Copyright 2024 Sony Semiconductor Israel, Inc. All rights reserved.
```
Comment (ofirgo) [inline model_compression_toolkit/core/pytorch/data_util.py:1]: If these function are only used as utils for the Hessian service then I think it's better to put it under "hessian" folder, to reduce the chance of using them for other purposes without paying attention

File: model_compression_toolkit/core/pytorch/data_util.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+            data_gen_fn: data generator factory.
+            n_samples: target size of the dataset. If None, use all samples yielded by the data generator in one pass.
+        """
+        test_batch = next(data_gen_fn())
+        if not isinstance(test_batch, list):
+            raise TypeError(f'Data generator is expected to yield a list of tensors, got {type(test_batch)}')
+        self.orig_batch_size = test_batch[0].shape[0]
+
+        samples = []
+        for batch in data_gen_fn():
```
Comment (ofirgo) [inline model_compression_toolkit/core/pytorch/data_util.py:93]: I can't locate this in the changes so I'll just ask - is this, or any of the other dataset generators defined here, run shuffle on the dataset? meaning, are the batches returned with different samples in each epoch or with the same samples in each batch?

Idealy, for GPTQ, we would want shuffling in the dataset

File: model_compression_toolkit/gptq/common/gptq_training.py [hunk 3]
```diff
@@ -261,11 +207,9 @@ def build_gptq_model(self):
...
     @abstractmethod
-    def train(self, representative_data_gen: Callable):
+    def train(self):
         """
-        Train the quantized model using GPTQ training process
-        Args:
-            representative_data_gen: Dataset to use for inputs of the models.
+        Train the quantized model using GPTQ training process.
         """
         raise NotImplemented(f'{self.__class__.__name__} have to implement the '
```
Comment (ofirgo) [inline model_compression_toolkit/gptq/common/gptq_training.py:214]: NotImplementedError? (;

File: model_compression_toolkit/gptq/pytorch/gptq_training.py [hunk 4]
```diff
@@ -70,6 +73,7 @@ def __init__(self,
             hessian_info_service: HessianInfoService to fetch info based on the hessian approximation of the float model.
         """
         def _get_total_grad_steps():
+            # TODO get it from the dataset
```
Comment (ofirgo) [inline model_compression_toolkit/gptq/pytorch/gptq_training.py:76]: It's not possible after this refactor? (probably not but I'm just making sure)

File: model_compression_toolkit/gptq/pytorch/gptq_training.py [hunk 5]
```diff
@@ -106,20 +111,84 @@ def _get_total_grad_steps():
...
+        hessians_tensor = torch.stack([layers_hessians[layer.name] for layer in self.compare_points], dim=1)    # samples X layers
+        assert hessians_tensor.shape[1] == len(self.compare_points)
+        loss_weights = list(hessians_tensor)
+        sla_train_dataset = FixedSampleInfoDataset(fixed_dataset.samples, loss_weights)
+
+        reg_weights = hessians_tensor.mean(dim=0)
+        # use collate to add a single value to each batch
+        collate_fn = get_collate_fn_with_extra_outputs(reg_weights)
+
+        return DataLoader(sla_train_dataset, batch_size=orig_batch_size, shuffle=True, collate_fn=collate_fn)
```
Comment (ofirgo) [inline model_compression_toolkit/gptq/pytorch/gptq_training.py:161]: I see now the "shuffle" flag so it probably answers my previous question regarding the shuffle, but I'm keeping it open just to make sure

**Review by ofirgo — COMMENTED** [review]
Still need to go over the tests but in the meanwhile leaving some comments

File: model_compression_toolkit/core/common/mixed_precision/sensitivity_evaluation.py [hunk 6]
```diff
@@ -238,22 +238,20 @@ def _compute_hessian_based_scores(self) -> np.ndarray:
         """
         # Create a request for Hessian approximation scores with specific configurations
         # (here we use per-tensor approximation of the Hessian's trace w.r.t the node's activations)
+        fw_dataloader = self.fw_impl.convert_data_gen_to_dataloader(self.representative_data_gen,
```
[the remark this answers is not in the document]
Comment (irenaby) [reply model_compression_toolkit/core/common/mixed_precision/sensitivity_evaluation.py:241]: Not necessarily. The request should contain the dataloader for hessian computation with the appropriate batch size and that's something that the caller must supply. It definitely shouldn't be a responsibility of the request to convert python data generators into dataloader. BTW dataloader is just defined as iterable, so if the batch size was appropriate data generator instance could be passed as is. Neither request not the hessian service care what exactly dataloader is.

File: model_compression_toolkit/core/common/pruning/importance_metrics/lfh_importance_metric.py [hunk 7]
```diff
@@ -120,22 +120,21 @@ def _get_entry_node_to_score(self, entry_nodes: List[BaseNode]) -> Dict[BaseNode
...
 
         # Initialize HessianInfoService for score computation.
+
         hessian_info_service = HessianInfoService(graph=self.float_graph,
-                                                  representative_dataset_gen=self.representative_data_gen,
                                                   fw_impl=self.fw_impl)
 
         # Fetch and process Hessian scores for output channels of entry nodes.
-        nodes_scores = []
-        for node in entry_nodes:
```
[the remark this answers is not in the document]
Comment (irenaby) [reply model_compression_toolkit/core/common/pruning/importance_metrics/lfh_importance_metric.py:129]: Yes, I though we talked about this, but maybe only I thought that :) I can restore it. Gradient-wise it should be identical but there is a small difference that now the same random vector is used for all nodes whereas previously random vector was generated for each node.

File: model_compression_toolkit/core/common/quantization/quantization_params_generation/error_functions.py [hunk 8]
```diff
@@ -377,23 +377,26 @@ def _get_sliced_histogram(bins: np.ndarray,
...
+                              num_hessian_samples: int,
+                              dataloader: Optional[Iterable]) -> List[List[np.ndarray]]:
     """
     Compute and retrieve Hessian-based scores for using during HMSE error computation.
 
     Args:
         node: The node to compute Hessian-based scores for.
         hessian_info_service: HessianInfoService object for retrieving Hessian-based scores.
         num_hessian_samples: Number of samples to approximate Hessian-based scores on.
+        dataloader: Data loader for computing Hessian-based scores. Can be None if hessians are expected to be available.
```
[the remark this answers is not in the document]
Comment (irenaby) [reply model_compression_toolkit/core/common/quantization/quantization_params_generation/error_functions.py:389]: It means they were already precomputed, I'll try to make it clearer

File: model_compression_toolkit/core/pytorch/data_util.py (hunk 1 above)
Comment (irenaby) [reply model_compression_toolkit/core/pytorch/data_util.py:1]: There is no reason not to use this for other purposes, as there is nothing hessian-specific here. That's exactly the reason it's here and not under hessian :)

File: model_compression_toolkit/core/pytorch/data_util.py (hunk 2 above)
Comment (irenaby) [reply model_compression_toolkit/core/pytorch/data_util.py:93]: For fixed dataset shuffling can be defined in dataloader as it's a regular torch Dataset. 
For iterable dataset shuffling is problematic and torch doesn't allow it, so the provided data generator should take care of that. This is exactly the same as today. Iterable dataset follows the data generator precisely, I think its mentioned in the documentation

File: model_compression_toolkit/gptq/common/gptq_training.py (hunk 3 above)
Comment (irenaby) [reply model_compression_toolkit/gptq/common/gptq_training.py:214]: I didn't write this line!

File: model_compression_toolkit/core/common/mixed_precision/sensitivity_evaluation.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ofirgo) [reply model_compression_toolkit/core/common/mixed_precision/sensitivity_evaluation.py:241]: Let me just verify that I understand correctly—the "user" (meaning us) can pass a Python generator, and it'll be okay?What do you mean by "if the batch size was appropriate"?
It just seems like complicated requirement for the user, that exposes internal implementation mechanism of the service and requires the user to deal with it.
If you say that the request excepts a "simple" python generator and here it is built as a loader just to define a different batch size then it's ok

File: model_compression_toolkit/core/common/pruning/importance_metrics/lfh_importance_metric.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ofirgo) [reply model_compression_toolkit/core/common/pruning/importance_metrics/lfh_importance_metric.py:129]: Yes you are correct that it should be identical, and also that the small change due to randomness is probably negligible, but I prefer to avoid changing unrelated algorithms to prevent unexpected issues, this is already a big enough refactor that affects many modules

File: model_compression_toolkit/gptq/common/gptq_training.py (hunk 3 above)
Comment (ofirgo) [reply model_compression_toolkit/gptq/common/gptq_training.py:214]: I know, but you also changed many other such exceptions in places you modified other code (like in the fw implementation) so I just wondered why here you decided to let it go 😄

File: tests/keras_tests/function_tests/test_hmse_error_method.py [hunk 9]
```diff
@@ -110,52 +109,58 @@ def _run_node_verification(node_type):
...
+                                                            data_loader=None,
+                                                            n_samples=1,
                                                             target_nodes=[node])
-
-            self.assertTrue(self.his.count_saved_scores_of_request(expected_hessian_request)[node] > 0,
-                            f"No Hessian-based scores were computed for node {node}, "
-                            "but expected parameters selection to run with HMSE.")
+            # check hessians have been precomputed (dataloader=None fetches from cache)
+            hess = self.his.fetch_hessian(expected_hessian_request)
+            self.assertTrue(hess[node.name].shape[0] == 1)
```
Comment (ofirgo) [inline tests/keras_tests/function_tests/test_hmse_error_method.py:119]: add a message to this assertion, so it would be simpler to understand what was the expected behavior (None dataloader should fetch from cache) in case it fails

File: tests/pytorch_tests/function_tests/test_hessian_info_calculator.py [hunk 10]
```diff
@@ -35,7 +36,7 @@
 class basic_model(torch.nn.Module):
     def __init__(self):
         super(basic_model, self).__init__()
-        self.conv1 = Conv2d(3, 3, kernel_size=1, stride=1)
+        self.conv1 = Conv2d(3, 3, kernel_size=3, stride=1, padding=1)
```
Comment (ofirgo) [inline tests/pytorch_tests/function_tests/test_hessian_info_calculator.py:39]: why is this change necessary?

**Review by ofirgo — CHANGES_REQUESTED** [review]

File: tests/pytorch_tests/function_tests/test_hessian_info_calculator.py (hunk 10 above)
Comment (irenaby) [reply tests/pytorch_tests/function_tests/test_hessian_info_calculator.py:39]: Small kernel with small input randomly zero the gradients (I assume relu randomly gets all-negative inputs). This was the simplest "solution" to significantly decrease the chance of that happening.

File: model_compression_toolkit/core/common/mixed_precision/sensitivity_evaluation.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (irenaby) [reply model_compression_toolkit/core/common/mixed_precision/sensitivity_evaluation.py:241]: Let's talk

File: model_compression_toolkit/gptq/common/gptq_training.py (hunk 3 above)
Comment (irenaby) [reply model_compression_toolkit/gptq/common/gptq_training.py:214]: I didn't decide, I just didn't notice it. Probably losing sight due to old age, that's the only possible explanation

Comment (ofirgo) [reply model_compression_toolkit/gptq/common/gptq_training.py:214]: that's what I feared of 😆

File: model_compression_toolkit/gptq/pytorch/gptq_training.py (hunk 4 above)
Comment (irenaby) [reply model_compression_toolkit/gptq/pytorch/gptq_training.py:76]: Not sure if there was an issue here or I just didn't want to waste time on something that currently works as I was instructed :)  Most probably it's the last one

File: model_compression_toolkit/core/common/hessian/hessian_info_service.py [hunk 11]
```diff
@@ -12,19 +12,91 @@
...
+        Returns:
+            Minimal samples count after update (among updated layers).
+
+        """
+        assert set(layers_hessians.keys()) == set(n.name for n in request.target_nodes)
+        n_nodes_samples = []   # samples count per node after update
+        for node_name, hess in layers_hessians.items():
+            query = Query(request.mode, request.granularity, node_name)
+            saved_hess = self._data.get(query)
+            new_hess = hess if saved_hess is None else np.unique(np.concatenate([saved_hess, hess], axis=0), axis=0)
```
[the remark this answers is not in the document]
Comment (irenaby) [reply model_compression_toolkit/core/common/hessian/hessian_info_service.py:60]: Yeah, calculators return the results as np. But you are right, I failed to keep it tensor-type agnostic, will change.

File: model_compression_toolkit/core/common/pruning/importance_metrics/lfh_importance_metric.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (irenaby) [reply model_compression_toolkit/core/common/pruning/importance_metrics/lfh_importance_metric.py:129]: restored

File: model_compression_toolkit/core/common/hessian/hessian_info_service.py [hunk 12]
```diff
@@ -12,21 +12,96 @@
...
+    granularity: HessianScoresGranularity
+    node: LayerName
+
+
+class HessianCache:
+    """ Hessian cache """
+    def __init__(self):
+        self._data: Dict[Query, Tensor] = {}
+
+    def update(self, layers_hessians, request) -> int:
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/hessian/hessian_info_service.py:46]: add typehints

**Review by ofirgo — APPROVED** [review]
